### PR TITLE
Add training script for upgrading to new versions of YOLO

### PR DESCRIPTION
**Pull request by speige**
This PR needs improvement before it should be considered for merging.

YOLO v11 has better accuracy with fewer parameters (smaller model file size) and faster inference than v8.
I wanted to upgrade to v11, so I made this script. 

It downloads your old training dataset from archive.org, labels each image (using the v8 NSFW model from github releases), then re-trains against v11.

Images are automatically renamed to .jpg_nsfw to avoid accidental trauma if user doesn't want to see images. Non-default file extension prevents operating system from rendering thumbnails/etc.

Specific Labels can be ignored during training if not important to user (face/feet/etc)

I'd like feedback on how to improve accuracy.

I believe you have a different training script that you've kept private? Are you willing to share that? Or, do you have any recommendations for improving this script?

I think the main improvement would come from having a better image dataset for training.

Your old v1 dataset from archive.org only contains 320x320 sized images, whereas your released model also has a 640x640 version. I suspect you've improved your web scraping since then and have a better private image dataset.

By the human eye my opinion is the old model is better. I think v11 is capable mathematically of better results, but training is only as good as your dataset and your tuning parameters. I compared the models with netron and believe I used the same tuning parameters as you.

I suspect my training dataset may have fewer images, be missing certain styles of NSFW, or may be labeled incorrectly, etc.

### CODE REVIEW

**Comment by PerfectFox265** [comment]
Can you upload the dataset on huggingface ? It takes days to download it from archive.org. I would like to improve it to differentiate male and female faces, and make it better on 2d art style.

You said it contains only images of 320x320. Are they stretched or have they black borders ? Also, why not just put the image in the top left corner instead of computing many padding-things to center it ?

**Comment by speige** [comment]
It downloads quickly for me, ~ 1 hour, not days. Try using a different internet connection or a VPN.
I didn't create the dataset, it was uploaded by the original author of NudeNet.

The newest version of NudeNet was definitely trained on a better dataset, but the author didn't make it publicly available, so this is the best dataset we have, but it's still not a great starting point.

The weird padding is called "letterboxing" as far as I know it's necessary for YOLO to work correctly. I think the NudeNet python & TypeScript libraries use this same technique. Top/Left instead of letterboxing will cause worse accuracy.

Also, this dataset is not labeled, so the labels are re-generated after downloading by running NudeNet against it. In order to improve the male/female face detection you'll need to re-label with a different technique.

There may be a simpler way to improve male/female face detection. Run a separate algorithm after detection which looks for FACE_MALE bounding boxes in close proximity to FEMALE_BREAST, and then swap FACE_MALE to FACE_FEMALE.

**Comment by PerfectFox265** [comment]
We could use [FaceONNX ](https://github.com/FaceONNX/FaceONNX/tree/main) to better classify genders. This can help for the labeling of the dataset. However, I think it's very hard to differentiate male/female faces for a full body image resized to 320x320 (even with human eyes). And nudenet also have trouble with male/female breast.

If I'm correct, setting imgsz to 640 would allow to use 640x640 tensors, do you know how much it slow down the detection ? I guess you aim for a yolov11-640, if you need to create a better dataset.

There are some nsfw datasets on hugging face, but I hope we can train a good model with less than 80GB of 640x640 images...

**Comment by speige** [comment]
Yes, you can change IMG_SIZE to 640

True, it does mis-classify gender of breast sometimes, but not nearly as often as faces.

NudeNet already has a 640x640 version, so we can use that to estimate before re-training. Unfortunately, this archive.org dataset has many images that are smaller than 640x640 so we may need a better dataset.

On my PC, running the same image against the 320x320 & 640x640 versions, both correctly classified all the labels, however 640x640 had a higher confidence of 96% instead of 89%. Both were extremely fast, but 320x320 was 5x faster (.025ms vs .165ms)

It all depends on your use case. For me, I'm using a quantized 320x320 because I care the most about speed and accuracy is good enough.

My reason to re-train NudeNet is to improve accuracy of "sexy" images (bra, bikini, etc), so I'd add more labels instead of breast_clothed and breast_exposed. I don't care about face or feet, so I'd remove those altogether for faster performance.

I tried ~ 5 different models before deciding on NudeNet. Most of the others are only a true/false for porn and are very inaccurate. NudeNet works much better for my needs.

I think 40GB or higher will be necessary. I think having a good dataset (wide range of image styles and accurate labels) is the most important. I think it can work well even with 320x320.

I think NudeNet didn't publish newer datasets because it generates them on the fly with a web scraper and deletes them afterwards. We could easily do the same. It shouldn't take long, the biggest issue is being careful not to accidentally download something illegal (use a VPN just in case), using enough unique terms to get different styles, and to label everything correctly.

Sometimes a classifier learns the wrong pattern even if it works most of the time. An example is male body builder vs female breast. To a human they're obviously different, but to an AI they're extremely similar. If we train on both styles, it should learn to distinguish, but if we don't train on both, by default it's going to think male body builders have female breasts. This is what I mean by training on "different styles". Similar examples would be anime, transparent clothing/wet t-shirt, etc.

I think accurately labeling the dataset will be the hardest part. Labeling 40GB by hand will take forever and cause severe trauma. We either need to label a small dataset by hand to train a base model and use that to label the large dataset. Or we need to use a combination of multiple existing models. At 1st glance, something like https://cloud.google.com/vision/docs/drag-and-drop should work (although it's paid), but when you test it on some images, you realize it only gives bounding boxes for certain things, like faces, but not breasts. Strange, because it still can detect it as "adult".

I don't personally want to be exposed to NSFW images if I can avoid it, so it makes development & debugging awkward. I rename images to .jpg_nsfw so the thumbnails don't render in windows.

**Comment by speige** [comment]
I realized I had a major bug in my code that was ruining accuracy, so I pushed some code changes and the results are much better. I think this is now a valid script for up-converting to different YOLO model architectures.

I think there's still plenty of room for improvement.
Many auto-labeling & training parameters could be tweaked
Labeling could be done better by using an ensemble of multiple models (example: FaceONNX for better male/female labeling)
Pose detection could be used to remove false positives from the auto-labeled dataset
Could code a web scraper to download higher resolution images, rather than using this ~ 320 imgsz dataset

**Comment by tien226anh** [comment]
> I realized I had a major bug in my code that was ruining accuracy, so I pushed some code changes and the results are much better. I think this is now a valid script for up-converting to different YOLO model architectures.
> 
> I think there's still plenty of room for improvement. Many auto-labeling & training parameters could be tweaked Labeling could be done better by using an ensemble of multiple models (example: FaceONNX for better male/female labeling) Pose detection could be used to remove false positives from the auto-labeled dataset Could code a web scraper to download higher resolution images, rather than using this ~ 320 imgsz dataset

Do you have any trained version of NudeNet by any chance?

**Comment by speige** [comment]
https://github.com/notAI-tech/NudeNet/releases
I'd recommend using these pre-trained models unless you want to train your own.
In my opinion they have the best accuracy of any free public model.

I'm working on a new model but it's not ready for public sharing yet.

**Comment by PerfectFox265** [comment]
I have uploaded my nudity detection models here : https://huggingface.co/Perfectfox256/hotscreen-detection-models/tree/main/yolo-07-2025
They are not perfect and use different classes than nudenet, but the yolo 11 S 640x640 is probably nice to help you labelize a dataset.
